dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/RelRMW.java

```diff
@@ -136,56 +102,53 @@ public TupleSet getMaxTupleSet(){
     protected BooleanFormula encodeApprox(SolverContext ctx) {
         FormulaManager fmgr = ctx.getFormulaManager();
 		BooleanFormulaManager bmgr = fmgr.getBooleanFormulaManager();
-        
-        // Encode base (not exclusive pairs) RMW
-        TupleSet origEncodeTupleSet = encodeTupleSet;
-        encodeTupleSet = new TupleSet(Sets.intersection(encodeTupleSet, baseMaxTupleSet));
-        BooleanFormula enc = super.encodeApprox(ctx);
-        encodeTupleSet = origEncodeTupleSet;
-
-        // Encode RMW for exclusive pairs
-		BooleanFormula unpredictable = bmgr.makeFalse();
-        for(Thread thread : task.getProgram().getThreads()) {
-            for (Event store : thread.getCache().getEvents(storeExclFilter)) {
-            	BooleanFormula storeExec = bmgr.makeFalse();
-                for (Event load : thread.getCache().getEvents(loadExclFilter)) {
-                    if (load.getCId() < store.getCId()) {
-
-                        // Encode if load and store form an exclusive pair
-                    	BooleanFormula isPair = exclPair(load, store, ctx);
-                    	BooleanFormula isExecPair = bmgr.and(isPair, store.exec());
-                        enc = bmgr.and(enc, bmgr.equivalence(isPair, pairingCond(thread, load, store, ctx)));
-
-                        // If load and store have the same address
-                        BooleanFormula sameAddress = generalEqual(((MemEvent)load).getMemAddressExpr(), ((MemEvent)store).getMemAddressExpr(), ctx);
-                        unpredictable = bmgr.or(unpredictable, bmgr.and(isExecPair, bmgr.not(sameAddress)));
-
-                        // Relation between exclusive load and store
-                        enc = bmgr.and(enc, bmgr.equivalence(this.getSMTVar(load, store, ctx), bmgr.and(isExecPair, sameAddress)));
-
-                        // Can be executed if addresses mismatch, but behaviour is "constrained unpredictable"
-                        // The implementation does not include all possible unpredictable cases: in case of address
-                        // mismatch, addresses of read and write are unknown, i.e. read and write can use any address
-                        storeExec = bmgr.or(storeExec, isPair);
-                    }
+        BooleanFormula enc = bmgr.makeTrue();
+        BooleanFormula unpredictable = bmgr.makeFalse();
+        Map<Event, BooleanFormula> map = new HashMap<>();
+        for(Tuple tuple : maxTupleSet) {
```

Comment:
Should be `encodeTupleSet` for all non-exclusive events. This is particularly important for Svcomp, because atomic blocks create huge `rmw`-relations of which not a single one needs to get encoded.
For exclusive events, I think you always need to encode the pairing conditions (not sure though).

Reply:
I slightly changed the semantics of the `excl` variables for this purpose:  It now accounts for the `MATCHADDRESS` tag of the store and is encoded for all non-min exclusive pairs, while the `rmw` relationship is defined on top of it.